src/extension/panel.ts

```diff
@@ -0,0 +1,162 @@
+            </body>
+            </html>`;
+
+        webview.onDidReceiveMessage(async message => {
+            if (!message) return;
+            switch (message.command) {
+            case 'open': {
+                const uris = await window.showOpenDialog({
+                    defaultUri: workspace.workspaceFolders?.[0]?.uri,
+                    filters: { 'SARIF files': ['sarif', 'json'] },
+                });
+                if (!uris) return;
+                store.logs.push(...await loadLogs(uris));
+                break;
+            }
+            case 'removeLog': {
+                store.logs.removeWhere(log => log._uri === message.uri);
+                break;
+            }
+            case 'select': {
+                const {logUri, uri, region} = message as { logUri: string, uri: string, region: _Region};
+                const validatedUri = await basing.translateArtifactToLocal(uri);
+                if (!validatedUri) return;
+                await this.selectLocal(logUri, validatedUri, region);
+                break;
+            }
+            case 'selectLog': {
+                const [logUri, runIndex, resultIndex] = message.id as ResultId;
+                const log = store.logs.find(log => log._uri === logUri);
+                const result = store.logs.find(log => log._uri === logUri)?.runs[runIndex]?.results?.[resultIndex];
+                if (!log || !result) return;
+                await this.selectLocal(logUri, log._uriUpgraded ?? log._uri, result._logRegion);
+                break;
+            }
+            case 'setState': {
+                const oldState = Store.globalState.get('view', defaultState);
+                const {state} = message;
+                await Store.globalState.update('view', Object.assign(oldState, JSON.parse(state)));
+                break;
+            }
+            default:
+                throw new Error(`Unhandled command: ${message.command}`,);
+            }
+        }, undefined, context.subscriptions);
+
+        await this.spliceLogs([], store.logs);
+    }
+
+    public async selectLocal(logUri: string, localUri: string, region: _Region | undefined) {
+        // Keep/pin active Log as needed
+        for (const editor of window.visibleTextEditors.slice()) {
+            if (editor.document.uri.toString() !== logUri) continue;
+            await window.showTextDocument(editor.document, editor.viewColumn);
+            await commands.executeCommand('workbench.action.keepEditor');
+        }
+
+        const doc = await workspace.openTextDocument(Uri.parse(localUri));
+        const editor = await window.showTextDocument(doc, ViewColumn.One, true);
+
+        if (region === undefined) return;
+        editor.selection = regionToSelection(doc, region);
+        editor.revealRange(editor.selection, TextEditorRevealType.InCenterIfOutsideViewport);
+    }
+
+    public select(result: Result) {
+        if (!result?._id) return; // Reduce Panel selection flicker.
```

Comment:
>!result?._id [](start = 12, length = 12)

so if _id is 0, this returns. That okay?

Reply:
_id is a ResultId which is `[string, number, number]`, so we should be safe here.

---
In reply to: [441650171](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441650171) [](ancestors = 441650171)